includes/admin/class-wcs-admin-post-types.php

```diff
@@ -660,14 +660,33 @@ public static function get_date_column_content( $subscription, $column ) {
 
 		$date_type_map = array( 'last_payment_date' => 'last_order_date_created' );
 		$date_type     = array_key_exists( $column, $date_type_map ) ? $date_type_map[ $column ] : $column;
+		$date_timestamp    = $subscription->get_time( $date_type, 'site' );
 
 		if ( 0 == $subscription->get_time( $date_type, 'gmt' ) ) {
 			$column_content = '-';
 		} else {
-			$column_content = sprintf( '<time class="%s" title="%s">%s</time>', esc_attr( $column ), esc_attr( date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $subscription->get_time( $date_type, 'site' ) ) ), esc_html( $subscription->get_date_to_display( $date_type ) ) );
+			$column_content = sprintf( '<time class="%s" title="%s">%s</time>', esc_attr( $column ), esc_attr( date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $date_timestamp ) ), esc_html( $subscription->get_date_to_display( $date_type ) ) );
 
-			if ( 'next_payment_date' == $column && $subscription->payment_method_supports( 'gateway_scheduled_payments' ) && ! $subscription->is_manual() && $subscription->has_status( 'active' ) ) {
-				$column_content .= '<div class="woocommerce-help-tip" data-tip="' . esc_attr__( 'This date should be treated as an estimate only. The payment gateway for this subscription controls when payments are processed.', 'woocommerce-subscriptions' ) . '"></div>';
+			// Custom handling for `Next payment` date column.
+			if ( 'next_payment_date' === $column ) {
+				$subscription_is_active = $subscription->has_status( 'active' );
+
+				$tooltip_message = '';
+				$tooltip_classes = 'woocommerce-help-tip';
+
+				if ( $subscription_is_active && $date_timestamp < current_time( 'timestamp', false ) ) {
```

Comment:
This is a phpcs warning, though it sounds possibly relevant, could be causing a bug.

>  677 | WARNING | [ ] Calling current_time() with a $type of "timestamp" or "U" is strongly discouraged as it will not return a Unix (UTC) timestamp. Please consider using a non-timestamp format or otherwise refactoring this code.
 
What I want to do here is compare the subscription last payment date (timestamp) with the current time (in timestamp format). In my testing this is working correctly (though maybe I missed a test case). 

What's the best way to compare dates (in WordPress/PHP) – what do we do elsewhere in the code?

And does that meet the phpcs standards?

I see that phpcs is stricter than our current code base, so whenever we touch code we'll need to address phpcs issues. Need to take care not to introduce bugs :)

Reply:
> What's the best way to compare dates (in WordPress/PHP) – what do we do elsewhere in the code?

The issue in this case is that the `$date_timestamp` is a timestamp in the site's timezone. 

I don't think we have any instance where we're trying to determine if a site-converted-timestamp is in the past. We are usually comparing a UTC/GMT one in which case we use either: 

- `time()`
- `gmdate( 'U' )` - GMT timestamp for the current time. 

---

I think there are 3 ways to resolve this:

1. Keep the timestamp variable `$date_timestamp` as a site timestamp. Then to determine if it's in the past you'll have to do some manipulation like this: `$date_timestamp < ( time() + wc_timezone_offset() )`. Which is basically offsetting the timestamp returned by `time()` so we're comparing like for like - two site timestamps.
2. Rather than trying to do both date-displaying (which is typically a date in the site's timezone) and date-logic (which is typically done in UTC) with the same date variable, you could split it. eg: 
   - use `$subscription->get_time( ... ) < time()` for logic conditions and 
   - use `$subscription->get_time( ..., 'site' )` above where it's displayed to the user.
3. Use the `WC_DateTime` object which has functions for easy access to both a site and UTC timestamps. I've included a couple of code snippets example below to illustrate how you could do it: 

Replace your `$date_timestamp = $subscription->get_time( $date_type, 'site' );` with this
```php
// Get a WC_Datetime instance - the timestamp passed in must be UTC. 
$datetime = wcs_get_datetime_from( $subscription->get_time( $date_type ) );
```

Replace the `date( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ), $date_timestamp )` above with the following

```php
// This will display the date in the site's timezone given the specified format.
$datetime->date_i18n( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ) );
```

Finally, replace this `$date_timestamp < current_time( 'timestamp', false )` condition with:

```php
// getTimestamp returns the UTC timestamp.
$datetime->getTimestamp() < time();
```

Reply:
Nice! TIL `DateTime` class, this is much clearer and more explicit.

```php
$column_content = sprintf( 
  '<time class="%s" title="%s">%s</time>', 
  esc_attr( $column ), 
  esc_attr( $datetime->date_i18n( __( 'Y/m/d g:i:s A', 'woocommerce-subscriptions' ) ) ), 
  esc_html( $subscription->get_date_to_display( $date_type ) ) 
);
```

Side note, I see we are rendering the date twice on the column - once for content, once for `title` attr. I'm tempted to split these out to variables to document why each format is slightly different. Do you recall why they are different?

Reply:
The title format is a raw date and the `get_date_to_display()` can return a relative date.

So the idea is you can hover the relative date and see the actual raw date. 

<img width="225" alt="Screen Shot 2023-01-25 at 9 33 33 am" src="https://user-images.githubusercontent.com/8490476/214444544-5cb4d300-5f8b-4880-8015-cf388b0bc1ec.png">

Reply:
Aha - similar to GitHub comment timestamps! I did a little refactor to capture that intention: b279a31a7519193fd1bbac32735124cc3953dcde